core/WFCORE-3435_capability_requirements_deployments.adoc

```diff
@@ -0,0 +1,184 @@
+This interface provides a `org.jboss.as.controller.CapabilityActivatorContext` interface that extends `org.jboss.msc.service.ServiceActivatorContext` and add a method to access the `org.jboss.as.controller.capability.CapabilityServiceSupport`.
+
+This `CapabilityServiceSupport` interface can then be used to integrate with capabilities.
+
+We also propose to make `CapabilityActivator` extends `ServiceActivator`. This is not strictly required but this let the user
+implement a single interface (and register it with Java `ServiceLoader`) instead of having to deal with two different interfaces depending on the requirements for capabilities in the user services.
+
+[source]
+----
++-----------------------------------+        +--------------------------------------+      +------------------------+
+|             <<MSC>>               |<|------|          << WildFly Core >>          |<|----|        << User >>      |
+|         ServiceActivator          |        |         CapabilityActivator          |      |    MyServiceActivator  |
++-----------------------------------+        +--------------------------------------+      +------------------------+
+| activate(ServiceActivatorContext) |        | activate(CapabilityActivatorContext) |
++-----------------------------------+        +--------------------------------------+
+        ^                                          ^
+       / \                                        / \
+        |                                          |
++---------------------------------------+      +---------------------------------------------------------+
+|              <<MSC>>                  |      |                 << WildFly Core >>                      |
+|       ServiceActivatorContext         |<|----|             CapabilityActivatorContext                  |
++---------------------------------------+      +---------------------------------------------------------+
+| getServiceTarget(): ServiceTarget     |      | getCapabilityServiceSupport(): CapabilityServiceSupport |
+| getServiceRegistry(): ServiceRegistry |      +---------------------------------------------------------+
++---------------------------------------+          ^
+                                                  / \
+                                                   |
+                                               +---------------------------------------------------------------+
+                                               |                 << WildFly Core >>                            |
+                                               |             CapabilityServiceSupport                          |
+                                               +---------------------------------------------------------------+
+                                               | hasCapability(name: String): boolean                          |
+                                               | getCapabilityServiceName(capabilityName: String): ServiceName |
```

Comment:
What if a capability is not backed by a service?

Reply:
This feature is to install an user service that depends on another service through the use of capabilities.
If the capability is not backed by a service, I'm not sure what the user service would be depending on.

Reply:
If the activated services are using the same interface which includes `getCapabilityRuntimeAPI` then I guess that works.  In this case the dependency is simply on the WildFly container existing, which is (AFAIK) always true when services are running.